// eslint-plugin-react-redux/dist/rules/mapDispatchToProps-prefer-parameters-names.js
'use strict';

var isReactReduxConnect = require('../isReactReduxConnect');

var argumentNames = ['dispatch', 'ownProps'];

var report = function report(context, node, i) {
  context.report({
    message: `mapDispatchToProps function parameter #${i} should be named ${argumentNames[i]}`,
    node
  });
};

var check = function check(context, params) {
  params.forEach(function (param, i) {
    if (argumentNames[i] && param.type !== 'ObjectPattern' && argumentNames[i] !== param.name) {
      report(context, param, i);
    }
  });
};

module.exports = function (context) {
  return {
    VariableDeclaration(node) {
      node.declarations.forEach(function (decl) {
        if (decl.id && decl.id.name === 'mapDispatchToProps') {
          if (decl.init && (decl.init.type === 'ArrowFunctionExpression' || decl.init.type === 'FunctionExpression')) {
            check(context, decl.init.params);
          }
        }
      });
    },
    FunctionDeclaration(node) {
      if (node.id && node.id.name === 'mapDispatchToProps') {
        check(context, node.params);
      }
    },
    CallExpression(node) {
      if (isReactReduxConnect(node)) {
        var mapDispatchToProps = node.arguments && node.arguments[1];
        if (mapDispatchToProps && (mapDispatchToProps.type === 'ArrowFunctionExpression' || mapDispatchToProps.type === 'FunctionExpression')) {
          check(context, mapDispatchToProps.params);
        }
      }
    }
  };
};
