// eslint-plugin-react-redux/dist/rules/mapStateToProps-prefer-hoisted.js
'use strict';

var utils = require('../utils');
var isReactReduxConnect = require('../isReactReduxConnect');

var report = function report(context, node) {
  context.report({
    message: 'constant arrays and objects should be initialized outside of mapStateToProps',
    node
  });
};

var isConstArrayOrObj = function isConstArrayOrObj(node, nested) {
  if (node && node.type === 'ObjectExpression') {
    return node.properties.reduce(function (acc, prop) {
      return acc && isConstArrayOrObj(prop.value, nested + 1);
    }, true);
  }
  if (node && node.type === 'ArrayExpression') {
    return node.elements.reduce(function (acc, el) {
      return acc && isConstArrayOrObj(el, nested + 1);
    }, true);
  }
  if (node && node.type === 'Literal' && nested > 0) {
    return true;
  }
  return false;
};

var checkProp = function checkProp(node, context) {
  if (isConstArrayOrObj(node, 0)) {
    report(context, node);
  }
};

var checkFunction = function checkFunction(context, body) {
  var returnNode = utils.getReturnNode(body);
  if (returnNode && returnNode.type === 'ObjectExpression') {
    returnNode.properties.forEach(function (prop) {
      return checkProp(prop.value, context);
    });
  }
};

module.exports = function (context) {
  return {
    VariableDeclaration(node) {
      node.declarations.forEach(function (decl) {
        if (decl.id && decl.id.name === 'mapStateToProps') {
          var body = decl.init.body;
          checkFunction(context, body);
        }
      });
    },
    FunctionDeclaration(node) {
      if (node.id && node.id.name === 'mapStateToProps') {
        checkFunction(context, node.body);
      }
    },
    CallExpression(node) {
      if (isReactReduxConnect(node)) {
        var mapStateToProps = node.arguments && node.arguments[0];
        if (mapStateToProps && mapStateToProps.body) {
          checkFunction(context, mapStateToProps.body);
        }
      }
    }
  };
};
